### ISSUE TITLE
Support Instance::of for configuration in DI

### ISSUE THREAD

**Issue by SamMousa** [yiisoft/yii2#14293]
When the DI container creates an object it first resolves the dependencies via the constructor.
After that, there's a fork.
1. The object implements `Configurable`, set the params as last argument for the constructor.
2. The object does not implement `Configurable`, set the params manually.

My proposal is to add support for resolving `Instance::of()` declarations here as well.

For example:
```php
Container::get($class, $params = [], $config = []) { ... }

$container->get('SimpleClass', [], ['db' => Instance::of('test')]);
```
This would allow for intelligent property injection in cases where the container is available.

My use case, I have an application purely configured via DI.
Suppose my DI configuration array looks like this: 
```php
...,
\yii\web\Application::class => [
    'components' => [
        'fileCache' => FileCache::class // Config via DI,
        'dbCache' => DbCache::class // Config via DI
    ]
    'modules' => [
        'moduleA1' => modules\ClassA::class,
        'moduleA2' => modules\ClassA::class

    ]
]
```
Note that both modules are backed by the same class so I cannot configure them (separately) via DI.
These modules can use a service that I'd like to configure. In this case it's a cache.
So I try to configure these modules in the application config.
```php
...,
\yii\web\Application::class => [
    'components' => [
        'fileCache' => FileCache::class // Config via DI,
        'dbCache' => DbCache::class // Config via DI
    ]
    'modules' => [
        'moduleA1' => [
            'class' => modules\ClassA::class,
            'cache' => FileCache::class
        ],
        'moduleA2' => modules\ClassA::class
    ]
]
```
Now in `moduleA1` the `public $cache = "FileCache"`. This means I have to go back to the DI container to get the instance. Instead I'd like to use this:
```php
...,
\yii\web\Application::class => [
    'components' => [
        'fileCache' => FileCache::class // Config via DI,
        'dbCache' => DbCache::class // Config via DI
    ]
    'modules' => [
        'moduleA1' => [
            'class' => modules\ClassA::class,
            'cache' => Instance::of(FileCache::class)
        ],
        'moduleA2' => modules\ClassA::class
    ]
]
```

I'm aware that I could provide the module definition / configuration as a closure to work around this, just think this might be a clean solution without any downsides.

Opinions please, I can make a PR if you like the idea.

**Comment by samdark** [comment]
Overall it looks logical to support it. I have no objections. @yiisoft/core-developers what do you think?

**Comment by cebe** [comment]
If a property supports being set by DI you ususally use code like this: https://github.com/yiisoft/yii2/blob/0973117de289d8676ec1dc40f68ab36279d7cd2d/framework/web/CacheSession.php#L59

Supporting `Instance::of()` should be okay in general but running an instanceof check on all configuration entries could slow down instantiation process.

**Comment by samdark** [comment]
Would be great to have some benchmarks indeed...

**Comment by SamMousa** [comment]
@cebe, that's not DI that's the service locator pattern.
I've run a simple benchmark for the instanceof check:
````
public function actionBenchmark() {
        $times = [];
        $count = 0;
        for($i = 0; $i < 100; $i++) {
            $instance = Instance::of('test');
            $s = microtime(true);
            for ($j = 0; $j < 100000; $j++) {
                if ($instance instanceof Instance) {
                    $count++;
                } else {
                    $count--;
                }
            }
            $e = microtime(true);
            $times[] = $e - $s;
        }
        echo '<pre>';
        echo "$i iterations of $j checks\n";
        echo "Average: " . number_format(array_sum($times) / count($times), 3) . "s\n";
        echo "Min: " . number_format(min($times), 3) . "s\n";
        echo "Max: " . number_format(max($times), 3) . "s\n";
        print_r($times);

    }
````

This shows that on average doing 100k of `instanceof` checks takes 0.004 seconds. 

I tried some actions in my application and they have about 300 total config parameters...

**Comment by samdark** [comment]
So there's no difference performance-wise. That's good.

**Comment by samdark** [comment]
@SamMousa do you have time for a pull request? We're going to release 2.0.14 soon so either it is in or not (we have it working in di package already).

**Comment by SamMousa** [comment]
I might have time tomorrow. Otherwise it'll have to wait (or someone else has to pick it up)

**Comment by samdark** [comment]
Seems to be fixed already by #15347 :)
